operator/pkg/reconciliation/decommission_node.go

```diff
@@ -0,0 +1,228 @@
+		maxReplicas := *statefulSet.Spec.Replicas
+		lastPodSuffix := stsLastPodSuffix(maxReplicas)
+
+		if maxReplicas > desiredNodeCount {
+			dcPatch := client.MergeFrom(dc.DeepCopy())
+			updated := false
+
+			updated = rc.setCondition(
+				api.NewDatacenterCondition(
+					api.DatacenterScalingDown, corev1.ConditionTrue)) || updated
+
+			if updated {
+				err := rc.Client.Status().Patch(rc.Ctx, dc, dcPatch)
+				if err != nil {
+					logger.Error(err, "error patching datacenter status for scaling down rack started")
+					return result.Error(err)
+				}
+			}
+
+			rc.ReqLogger.Info(
+				"Need to update the rack's node count",
+				"Rack", rackInfo.RackName,
+				"maxReplicas", maxReplicas,
+				"desiredSize", desiredNodeCount,
+			)
+
+			rc.Recorder.Eventf(rc.Datacenter, corev1.EventTypeNormal, events.ScalingDownRack,
+				"Scaling down rack %s", rackInfo.RackName)
+
+			if err := setOperatorProgressStatus(rc, api.ProgressUpdating); err != nil {
+				return result.Error(err)
+			}
+
+			err := rc.DecommissionNodeOnRack(rackInfo.RackName, lastPodSuffix)
+			if err != nil {
+				return result.Error(err)
+			}
+
+			return result.RequeueSoon(10)
+		}
+	}
+
+	return result.Continue()
+}
+
+func (rc *ReconciliationContext) DecommissionNodeOnRack(rackName string, lastPodSuffix string) error {
+	for _, pod := range rc.dcPods {
+		podRack := pod.Labels[api.RackLabel]
+		if podRack == rackName && strings.HasSuffix(pod.Name, lastPodSuffix) {
+			mgmtApiUp := isMgmtApiRunning(pod)
+			if !mgmtApiUp {
+				return fmt.Errorf("Management API is not up on node that we are trying to decommission")
+			}
+
+			if err := rc.NodeMgmtClient.CallDecommissionNodeEndpoint(pod); err != nil {
+				// TODO this returns a 500 when it works
+				// We are waiting for a new version of mgmt api with a fix for this
+				// return false, err
+			}
+
+			rc.ReqLogger.Info("Marking node as decommissioning")
+			patch := client.MergeFrom(pod.DeepCopy())
+			pod.Labels[api.CassNodeState] = stateDecommissioning
+			if err := rc.Client.Patch(rc.Ctx, pod, patch); err != nil {
+				return err
+			}
+
+			rc.Recorder.Eventf(rc.Datacenter, corev1.EventTypeNormal, events.LabeledPodAsDecommissioning,
+				"Labeled node as decommissioning %s", pod.Name)
+
+			return nil
+		}
+	}
+
+	// this shouldn't happen
+	return fmt.Errorf("Could not find pod to decommission on rack %s", rackName)
+}
+
+// Wait for decommissioning nodes to finish before continuing to reconcile
+func (rc *ReconciliationContext) CheckDecommissioningNodes(epData httphelper.CassMetadataEndpoints) result.ReconcileResult {
+	for _, pod := range rc.dcPods {
+		if pod.Labels[api.CassNodeState] == stateDecommissioning {
+			if !IsDoneDecommissioning(pod, epData) {
+				rc.ReqLogger.Info("Node decommissioning, reconciling again soon")
+			} else {
+				rc.ReqLogger.Info("Node finished decommissioning")
+				rc.ReqLogger.Info("Deleting pod PVCs")
+
+				err := rc.DeletePodPvcs(pod)
```

Comment:
So most of the time, PVC have a finalizer on them that prevents them from being deleted while attached to a pod, but it looks like we delete the PVC while the pod is still around. Did that not cause issues in your testing? I think Ryan just recently had that issue. We could find the PVCs after scaling down the statefulset and the pods being terminated, or remove the the finalizers to delete the PVC. Also, I take it that when a pod is decommissioned, the management API continues running so the pod doesn't automatically terminate?

Reply:
Actually, maybe nothing to do here since you have a working test.

Reply:
Yeah I didn't hit any issues with my testing, but that could definitely be a concern. 

I chose to delete the PVC first because I was worried about a scenario in which we delete the pod first but then somehow fail to delete the PVC, we would never catch it on the next reconcile loop because there would be no pod left to perform a PVC lookup against and it would be orphaned out there until someone caught it.

Although I suppose we could potentially just look up all PVCs that exist and compare them against our current pods and see if there are any anomalies for that case, but we would need to do that check on every single reconcile loop (maybe that's OK?)

Reply:
To be clear, I'm fine with merging as things are. However, looking at the PVC:

```
  labels:
    app.kubernetes.io/managed-by: cass-operator
    cassandra.datastax.com/cluster: cluster1
    cassandra.datastax.com/datacenter: dc1
    cassandra.datastax.com/rack: r2
  name: server-data-cluster1-dc1-r2-sts-0
``` 

Now I wouldn't trust the "managed-by" label in this case, as in 1.0 the value of this label was wrong and couldn't be fixed for PVCs belonging DCs that already existed, but what you could do is get all PVCs for the DC and use the name to determine if the PVC should be removed. You don't have to do this all the time, just when your condition ScalingDown is true.

I know you have other work related to this, so maybe this is just something to think about for now and potentially address in one of your other tickets.

Reply:
Ah right, the "scalingDown" condition would still be true even if we failed to delete the pvc during the decomm loop, so we would know to do the check next time. That shouldn't be too bad then.

Reply:
Actually, I've been thinking about this more, and I think you have the right of it. I was thinking of `kubectl` which will block until the finalizers run, but I don't think that's the case with the go-client? In which case, you have this right. You delete the PVC, the finalizer prevents the PVC from going away, but then you scale down and the statefulset controller removes the pod, which removes the finalizer from the PVC, which allows the PVC to be garbage collected. Sorry for the confusion, what you have should be correct.

You _might_ need to set DeletePropagation (https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#DeletionPropagation) to DeletePropagationBackground, but if it is currently working, then you are probably fine.